File: physsynth/core/banded.py
"""Banded Cholesky — the one solver the theta-scheme string family shares.

Four models factor a pentadiagonal SPD update matrix and back-substitute every timestep:
:mod:`~physsynth.core.string_stiff`, :mod:`~physsynth.core.string_damped`,
:mod:`~physsynth.core.string_nonlinear` and :mod:`~physsynth.core.string_geometric`. Until this
module existed each of them imported ``cholesky_banded`` / ``cho_solve_banded`` straight from SciPy,
which was fine while there was only one implementation and is exactly wrong once there are two.

**Why this module exists at all** (``docs/dev/rust-migration-plan.md`` section 15). Those four
models are chained together by *bit-identity* reduction anchors -- ``sigma1 = 0`` makes the damped
string ``array_equal`` to the stiff one, ``EA = 0`` makes the tension-modulated string
``array_equal`` to the damped one, ``EA = T`` makes the geometrically exact string ``array_equal``
to it as well. Those tests compare one model against another, not against a closed form, so they
hold only while every model in the chain does the *same arithmetic in the same order*. Swapping any
single model to Rust would turn an intra-Python anchor into a cross-language one, and the banded
solve cannot carry that: OpenBLAS's triangular band kernel is blocked, and no scalar transcription
reproduces it (measured -- see the plan). Porting the **solver** instead of a model keeps all four
in step, because all four call this.

This is Phase 1's move one level down: :mod:`~physsynth.core.operators` is not a model either, and
flipping it swung five models at once.

Headless: NumPy + SciPy. No I/O, no plotting.
"""

from __future__ import annotations

import os

import numpy as np
from numpy.typing import NDArray
from scipy.linalg import cho_solve_banded, cholesky_banded

__all__ = ["cho_solve_banded", "cholesky_banded"]


# --- the Rust swap (docs/dev/rust-migration-plan.md, Phase 3) -----------------------------------
#
# The reference implementation here is SciPy's, i.e. LAPACK's -- unlike every other swap in this
# migration, the `_py` alias is not a Python transcription but the library call the acceptance
# numbers came from. That is the honest labelling: what this module ports is the *choice of
# solver*, and the Python side of the comparison is dpbtrf/dpbtrs.
#
# What the Rust side is, precisely: the reference LAPACK algorithms transcribed plainly -- DPBTF2
# for the factor (which is the path dpbtrf takes at kd = 2, since NB > KD), DTBSV twice for the
# solve. Two details are not cosmetic and are written down where they live, in
# `crates/physsynth-core/src/banded.rs`: DSCAL forms the reciprocal ONCE and multiplies, which is
# the reference algorithm's own behaviour and worth 19/120 -> 120/120 agreement on this family's
# matrices; and the fused multiply-add in DSYR is deliberately NOT reproduced, because fusing is a
# property of the kernel OpenBLAS picks at run time and asserting it would be a claim about a CPU.
#
# So this swap CHANGES THE NUMBERS. It is the first one in the migration that does so by design.
# What it does not change is any physics bar: measured on 2026-08-27, the transcribed solver's
# lossless energy drift is 2.7e-12 against LAPACK's 2.7e-12 on the same string (the bar is 1e-10),
# and the two trajectories agree to ~1e-13 of amplitude over a short run, growing slowly. The four
# models' agreement with each other stays EXACT, which is the property the anchors need.
#
# Off by default. SciPy is still the reference oracle.
cholesky_banded_py = cholesky_banded
cho_solve_banded_py = cho_solve_banded

_USE_RUST = os.environ.get("PHYSSYNTH_RS", "").strip() not in ("", "0", "false", "False")

if _USE_RUST:  # pragma: no cover - exercised by the dedicated CI job, not the default gate
    import physsynth_rs as _rs
    from scipy.linalg import LinAlgError

    def cholesky_banded(  # type: ignore[misc]  # noqa: F811
        ab: NDArray[np.float64],
        overwrite_ab: bool = False,
        lower: bool = False,
        check_finite: bool = True,
    ) -> NDArray[np.float64]:
        """``scipy.linalg.cholesky_banded``, on the Rust transcription of DPBTF2.

        ``overwrite_ab`` is accepted and ignored -- the Rust side always returns a fresh array,
        which is what every caller in this project already assumes. So is ``check_finite``: the
        Rust side always checks, because it does so in the pass that copies the input and a
        separate ``np.isfinite`` here cost more than the whole port saved. ``lower=True`` is
        refused rather than silently transposed: no call site asks for it, so it would be
        untested code.
        """
        if lower:
            raise NotImplementedError(
                "the Rust banded Cholesky implements upper storage only; every call site in "
                "this project passes lower=False"
            )
        try:
            return _rs.cholesky_banded_upper(ab)
        except _rs.NotPositiveDefinite as exc:
            raise LinAlgError(str(exc)) from None

    def cho_solve_banded(  # type: ignore[misc]  # noqa: F811
        cb_and_lower: tuple[NDArray[np.float64], bool],
        b: NDArray[np.float64],
        overwrite_b: bool = False,
        check_finite: bool = True,
    ) -> NDArray[np.float64]:
        """``scipy.linalg.cho_solve_banded``, on the Rust transcription of DPBTRS."""
        cb, lower = cb_and_lower
        if lower:
            raise NotImplementedError(
                "the Rust banded solve implements upper storage only; every call site in this "
                "project passes lower=False"
            )
        return _rs.cho_solve_banded_upper(cb, b)

File: crates/physsynth-core/src/banded.rs
//! Banded Cholesky — the shared solver behind the θ-scheme string family.
//!
//! `docs/dev/rust-migration-plan.md` §15. This module is the Phase 3 analogue of Phase 1's `ops`:
//! **not a model, but what four models are built out of.** `string_stiff`, `string_damped`,
//! `string_nonlinear` and `string_geometric` all factor a pentadiagonal SPD update matrix once (or,
//! in `string_nonlinear`'s case, once per distinct tension increment) and back-substitute every
//! timestep. Porting the solver before any of its callers is what keeps the family's four
//! bit-identity reduction anchors alive — see §15.2, which is the finding that forced this shape.
//!
//! # What is transcribed, and what deliberately is not
//!
//! SciPy dispatches `cholesky_banded` to LAPACK `dpbtrf` and `cho_solve_banded` to `dpbtrs`. At
//! `kd = 2` the blocked path in `dpbtrf` is never taken (`NB > KD`), so the factor is the
//! unblocked `DPBTF2`, and `dpbtrs` is two `DTBSV` calls. Both are transcribed here in their
//! reference form, with one detail that is *not* a detail:
//!
//! * `DPBTF2` scales the off-diagonal column through `DSCAL`, which forms the reciprocal **once**
//!   and multiplies. `x * (1/ajj)` is not `x / ajj` in binary floating point, and measured against
//!   OpenBLAS on 2026-08-27 the reciprocal spelling agreed on 120/120 of this family's matrices
//!   where the division spelling agreed on 19/120. That is the reference algorithm's own
//!   behaviour, not a property of a kernel, so it is transcribed.
//!
//! What is **not** transcribed is the fused multiply-add. With `fma` in the rank-1 update the
//! factor reproduces OpenBLAS exactly (120/120 against 82/120 without), but fusing is a property
//! of the `DSYR` kernel that `DYNAMIC_ARCH` picks at run time — asserting it would be a claim
//! about a CPU rather than about a port, which is the line §14.2 drew and this module keeps.
//!
//! The solve is not reproducible at all, and that was measured rather than assumed: per-element
//! candidate elimination over {forward, reverse} × {plain, fused} × {divide, reciprocal} on a
//! 15-element system left *disjoint* candidate sets — element 2 admitted only the dividing forms,
//! element 5 only the fused ones, element 9 only the forward order, element 14 excluded it — and
//! element 7 admitted nothing at all. No scalar recurrence produces OpenBLAS's `DTBSV`; it is a
//! blocked kernel. So this module does not chase it. It sums plainly, left to right, and the
//! family's agreement with LAPACK becomes a tolerance rather than an identity (§15.3).
//!
//! # Layout
//!
//! `ab` is row-major `(kd + 1) × n`, which is what a C-contiguous NumPy array of that shape hands
//! over, and it is *upper* banded storage: `ab[kd + i - j][j] == A[i][j]` for `j - kd <= i <= j`.
//! Row `kd` is the diagonal. Only the upper form is implemented — every call site in the project
//! passes `lower=False`, and a lower variant nothing calls would be untested code.

/// Why a factorization was refused.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BandedError {
    /// `A` is not positive definite; the 1-based index is the leading minor that failed, which is
    /// the number LAPACK returns in `info` and SciPy quotes back in its `LinAlgError`.
    NotPositiveDefinite(usize),
    /// The array's shape does not describe a band: zero columns, or a length that is not a
    /// multiple of `n`.
    BadShape,
}

impl std::fmt::Display for BandedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            BandedError::NotPositiveDefinite(i) => {
                write!(f, "{i}-th leading minor not positive definite")
            }
            BandedError::BadShape => write!(f, "ab must be a (kd + 1, n) band with n >= 1"),
        }
    }
}

/// Index into row-major `(kd + 1) × n` storage.
#[inline]
fn at(n: usize, row: usize, col: usize) -> usize {
    row * n + col
}

/// `A = U^T U`, upper banded — LAPACK `DPBTF2`, `UPLO = 'U'`.
///
/// `ab` is consumed by value and returned factored, mirroring `cholesky_banded`'s contract of
/// handing back a new array of the same shape rather than aliasing the input.
pub fn cholesky_banded_upper(
    mut ab: Vec<f64>,
    kd: usize,
    n: usize,
) -> Result<Vec<f64>, BandedError> {
    if n == 0 || ab.len() != (kd + 1) * n {
        return Err(BandedError::BadShape);
    }
    for j in 0..n {
        let mut ajj = ab[at(n, kd, j)];
        // `!(ajj > 0.0)`, not `ajj <= 0.0`: a NaN diagonal compares false against both, and it is
        // the negated form that then refuses. DPBTF2 tests `AJJ.LE.ZERO` on a value it has already
        // guaranteed non-NaN; here the caller may not have, so the refusal has to catch it.
        #[allow(clippy::neg_cmp_op_on_partial_ord)]
        if !(ajj > 0.0) {
            return Err(BandedError::NotPositiveDefinite(j + 1));
        }
        ajj = ajj.sqrt();
        ab[at(n, kd, j)] = ajj;

        // The trailing part of this column that still fits inside the band.
        let kn = kd.min(n - 1 - j);
        if kn == 0 {
            continue;
        }

        // DSCAL(kn, 1/ajj, ...) — reciprocal formed ONCE, then multiplied. See the header.
        let inv = 1.0 / ajj;
        for i in 0..kn {
            ab[at(n, kd - 1 - i, j + 1 + i)] *= inv;
        }

        // DSYR('U', kn, -1, x, ...) — the rank-1 update of the trailing block, where
        // `x[i] == A[j][j + 1 + i]` is the column just scaled. `alpha` multiplies `x[q]` first,
        // which is the order the reference kernel writes and therefore the order of the rounding.
        //
        // The read locations (row `kd - 1 - i`) and the write locations (row `kd + p - q`, with
        // `p <= q`) can never coincide, so this reads its own output only where LAPACK does too.
        for q in 0..kn {
            // `-1.0 * x` rather than `-x`, because DSYR's `alpha` multiplies `x[q]` first and the
            // rounding of this batch's transcription is the point of it. Negation is exact for a
            // finite double so the two agree — but the spelling is the reference kernel's.
            #[allow(clippy::neg_multiply)]
            let temp = -1.0 * ab[at(n, kd - 1 - q, j + 1 + q)];
            for p in 0..=q {
                let xp = ab[at(n, kd - 1 - p, j + 1 + p)];
                ab[at(n, kd + p - q, j + 1 + q)] += xp * temp;
            }
        }
    }
    Ok(ab)
}

/// `x := inv(U^T) x` — LAPACK `DTBSV('U', 'T', 'N')`, `incx = 1`.
fn tbsv_upper_transpose(chol: &[f64], kd: usize, n: usize, x: &mut [f64]) {
    for j in 0..n {
        let lo = j.saturating_sub(kd);
        let mut t = x[j];
        for i in lo..j {
            t -= chol[at(n, kd - (j - i), j)] * x[i];
        }
        x[j] = t / chol[at(n, kd, j)];
    }
}

/// `x := inv(U) x` — LAPACK `DTBSV('U', 'N', 'N')`, `incx = 1`.
///
/// The `x[j] != 0` guard is the reference kernel's, kept rather than simplified away: skipping the
/// update leaves a stored `-0.0` alone where subtracting `t * a` would flip its sign, and a signed
/// zero in the state is the kind of difference that survives to a comparison and explains nothing.
fn tbsv_upper_notranspose(chol: &[f64], kd: usize, n: usize, x: &mut [f64]) {
    for j in (0..n).rev() {
        if x[j] != 0.0 {
            x[j] /= chol[at(n, kd, j)];
            let t = x[j];
            let lo = j.saturating_sub(kd);
            for i in (lo..j).rev() {
                x[i] -= t * chol[at(n, kd - (j - i), j)];
            }
        }
    }
}

/// Solve `A x = b` given `A`'s upper banded Cholesky factor — LAPACK `DPBTRS`, `UPLO = 'U'`.
pub fn cho_solve_banded_upper(
    chol: &[f64],
    kd: usize,
    n: usize,
    b: &[f64],
) -> Result<Vec<f64>, BandedError> {
    if n == 0 || chol.len() != (kd + 1) * n || b.len() != n {
        return Err(BandedError::BadShape);
    }
    let mut x = b.to_vec();
    tbsv_upper_transpose(chol, kd, n, &mut x);
    tbsv_upper_notranspose(chol, kd, n, &mut x);
    Ok(x)
}
